lib/src/decorator.dart

```diff
@@ -0,0 +1,150 @@
+// limitations under the License.
+import 'dart:html';
+
+import 'package:meta/meta.dart';
+import 'package:platform_detect/platform_detect.dart';
+import 'package:platform_detect/src/support.dart';
+
+/// The CSS class that will be used to indicate that [decorateRootNodeWithPlatformClasses]
+/// has already been invoked for a given root node.
+const String _decorationCompleteClassName = 'pd--decorated';
+
+/// The CSS class that will be used to represent the current [browser].
+final String _browserClassName =
+    browser.className ?? nameToClassName(browser.name);
+
+/// The CSS class that will be used to represent the current [operatingSystem].
+final String _osClassName = nameToClassName(operatingSystem.name);
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [_browserClassName].
+const String _browserClassPrefix = 'ua-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [_osClassName].
+const String _osClassPrefix = 'os-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [_osClassName].
+const String _versionRangeClassPrefix = 'lt-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [Feature.name] if [Feature.isSupported] is `false`.
+const String _featureSupportNegationClassPrefix = 'no-';
+
+/// The first major release after the current one.
+@visibleForTesting
+final int minVersionRange = browser.version.major + 1;
+
+/// The number of major releases above [minVersionRange]
+/// that [_getBrowserVersionClasses] should produce CSS classes for.
+@visibleForTesting
+const int maxVersionRange = 2;
+
+String _getBrowserVersionClasses() {
+  var majorVersion = browser.version.major;
+  var classes = [
+    _browserClassPrefix + _browserClassName + majorVersion.toString()
+  ];
+
+  for (var i = minVersionRange; i < minVersionRange + maxVersionRange; i++) {
+    classes.add(
+        '$_browserClassPrefix$_versionRangeClassPrefix$_browserClassName$i');
+  }
+
+  return _listToClassNameString(classes);
+}
+
+String _getFeatureSupportClasses(List<Feature> features) {
+  var classes = [];
+
+  for (var i = 0; i < features.length; i++) {
+    var feature = features[i];
+
+    classes.add(feature.isSupported
+        ? feature.name
+        : '${_featureSupportNegationClassPrefix}${feature.name}');
+  }
+
+  return _listToClassNameString(classes);
+}
+
+/// Utility fn that takes a list of [classes] and returns a space-separated string for
+/// use within the `class` attribute on the `<html>` element that gets injected by
+/// [decorateRootNodeWithPlatformClasses].
+String _listToClassNameString(List<String> classes) {
+  var classesStr = '';
+
+  for (var i = 0; i < classes.length; i++) {
+    if (classesStr.isEmpty) {
+      classesStr = classes[i];
+    } else {
+      classesStr += ' ${classes[i]}';
+    }
+  }
+
+  return classesStr;
+}
```

Comment:
#nit This could be simplified:
```dart
String _listToClassNameString(List<String> classes) =>
    classes.where((classStr) => classStr.isNotEmpty).join(' ');
```